### PR TITLE
don't train or untrain empty word hashes

### PR DESCRIPTION
**Pull request by Ch4s3**

### CODE REVIEW

File: lib/classifier-reborn/bayes.rb [hunk 1]
```diff
@@ -73,7 +73,9 @@ def train(category, text)
 
       @backend.update_category_training_count(category, 1)
       @backend.update_total_trainings(1)
-      Hasher.word_hash(text, @language, @enable_stemmer).each do |word, count|
+      word_hash = Hasher.word_hash(text, @language, @enable_stemmer)
+      return if word_hash.length == 0
```
Comment (ibnesayeed) [inline lib/classifier-reborn/bayes.rb]: I would make it the very first thing to happen in the method. This current change is essentially changing nothing significantly. `update_category_training_count` and `update_total_trainings` are still being called, hence causing undesired stats changes. Also, if the `@auto_categorize` is set to `true`, it might add a new category without any trainings in it.

File: lib/classifier-reborn/bayes.rb [hunk 2]
```diff
@@ -91,7 +93,9 @@ def untrain(category, text)
       category = CategoryNamer.prepare_name(category)
       @backend.update_category_training_count(category, -1)
       @backend.update_total_trainings(-1)
-      Hasher.word_hash(text, @language, @enable_stemmer).each do |word, count|
+      word_hash = Hasher.word_hash(text, @language, @enable_stemmer)
+      return if word_hash.length == 0
```
Comment (ibnesayeed) [inline lib/classifier-reborn/bayes.rb]: Move this up in the method as well.

File: lib/classifier-reborn/bayes.rb (hunk 1 above)
Comment (Ch4s3) [inline lib/classifier-reborn/bayes.rb]: Do you want to preempt `add_category` or just `update_category_training_count` and `update_total_trainings`?

**Comment by ibnesayeed** [comment]
Also, using `word_hash.empty?` instead of `word_hash.length == 0` would be more elegant in my opinion.

**Comment by ibnesayeed** [comment]
Additionally, the `classifications` method should return a score hash with `Infinity` value for each category if the `word_hash` is empty like this:

```ruby
def classifications(text)
  score = {}
  word_hash = Hasher.word_hash(text, @language, @enable_stemmer)
  if word_hash.empty
    category_keys.each do |category|
      score[category.to_s] = Float::INFINITY
    end
    return score
  end
  #... remaining code
end

File: lib/classifier-reborn/bayes.rb (hunk 1 above)
Comment (ibnesayeed) [inline lib/classifier-reborn/bayes.rb]: I would make it the very first thing in the method, even before calling `CategoryNamer`. This way, if the training has to be skipped, it does not make any changes in the data-structure state in any circumstances.

**Comment by ibnesayeed** [comment]
I had these changes in a local branch on my code base, but since you are taking care of it, I will just discard that.

**Comment by Ch4s3** [comment]
So I made these changes, but it breaks [this test](https://github.com/jekyll/classifier-reborn/blob/master/test/bayes/bayesian_common_tests.rb#L89) .

```ruby
10.times do |a_number|
      assert_equal 'Digit', b.classify(a_number.to_s)
    end
```

But, I'm not sure it even makes sense to classify a single digit that way.

File: lib/classifier-reborn/bayes.rb [hunk 3]
```diff
@@ -112,18 +116,25 @@ def untrain(category, text)
...
+            score[category.to_s] += Math.log(s / total)
+          end
+          # now add prior probability for the category
+          s = @backend.category_has_trainings?(category) ? @backend.category_training_count(category) : 0.1
+          score[category.to_s] += Math.log(s / @backend.total_trainings.to_f)
         end
-        # now add prior probability for the category
-        s = @backend.category_has_trainings?(category) ? @backend.category_training_count(category) : 0.1
-        score[category.to_s] += Math.log(s / @backend.total_trainings.to_f)
+        score
```
Comment (ibnesayeed) [inline lib/classifier-reborn/bayes.rb]: Since you did not used `return` keyword in the `if` section of the `word_hash.empty?` clause and wrapped the remaining code in the `else` clause, I would encourage to remove implicit return of `score` from both the clauses and return it outside the conditional as the last line.

Alternatively, explicit return in the special case and not wrapping rest of the code in `else` clause would reduce a level of nesting though, while gaining the same effect.

**Comment by ibnesayeed** [comment]
> So I made these changes, but it breaks this test .

This is happening because how the tokenization is implemented currently. If the length of the token is not greater than 3 letters then it is skipped and not added in the `word_hash`. I think it was a meaningless test because there was only one category present and the system was returning that category with some score no matter what was passed to classify. With empty strings getting `Infinity` score, and a threshold is set, the `classify` method would return `nil`. Those categories were actually not being trained with anything.

```ruby
next unless word.length > 2 && !STOPWORDS[language].include?(word)
```

In conclusion, we need to change the test so that it trains for numbers above 100. This should make the test pass. Then we need to think if we want the limit of length where the token is nothing but a number such as `2`, `20`, or `.5`. This makes it more important to be able to pass a custom tokenizer (#131) for situations like this.

**Comment by Ch4s3** [comment]
I changed the test to use words, which I think better reflects the intent, and I tweaked the threshold. But in general I agree that it was a bad test, we seem to have a lot of those.

Yeah, let's prioritize #131, it will be pretty useful going forward.

**Comment by Ch4s3** [comment]
@ibnesayeed any insight into why `bayesian_integration_test.rb:35` fails on newer rubies with this PR?

**Comment by ibnesayeed** [comment]
> I changed the test to use words, which I think better reflects the intent...

You might want to make it something like `100.upto(110) do ...`.

**Comment by Ch4s3** [comment]
@ibnesayeed thoughts on the failing test?

**Comment by ibnesayeed** [comment]
> @ibnesayeed any insight into why bayesian_integration_test.rb:35 fails on newer rubies with this PR?

I am trying to look into it. If it was failing for all Rubies, I would have though there might be some edge condition on how numbers are stored in Redis.

**Comment by Ch4s3** [comment]
It might be an issue in the Redis Ruby gem

**Comment by ibnesayeed** [comment]
I think the issue might be on another level. Because even the expected value changes of each run. That expected value is calculated by generating a hash digest of the array of the results and scores concatenated. If the array is somehow changing the order or the digest calculation is not reliable to test the equality then we might need to make some changes there.

**Comment by ibnesayeed** [comment]
I think I got the culprit, just need to dig deeper and see how it goes.

**Comment by Ch4s3** [comment]
@ibnesayeed good catch

**Comment by ibnesayeed** [comment]
The issue is related to how `classify_with_score` method returns the value.

```ruby
def classify_with_score(text)
  (classifications(text).sort_by { |a| -a[1] })[0]
end
```

This returns the first value with the highest score. However, if more than one categories have the same score, how to resolve the conflict? In this particular test case, there is an instance in the dataset on like 4499, there is a record `ham	Ok` which is being used as test. Due to the only-stopword string, both teh categories are assigned `Infinity` value, and the method might return one or the other as the chosen one.

**Comment by ibnesayeed** [comment]
If the test uses a threshold, the situation will be little different as discussed in #123.

**Comment by ibnesayeed** [comment]
However, should have a fix for this situation without making any changes in the library code, only in the test. I will let you know those changes in a bit.

**Comment by Ch4s3** [comment]
awesome, thanks!

**Comment by ibnesayeed** [comment]
Please cherry pick https://github.com/ibnesayeed/classifier-reborn/pull/1/commits/ed116ea42aeeaa736d6245c483ae91ea25663a99

**Comment by ibnesayeed** [comment]
I was in rush and could not make proper PR on time. PR #133 should take care of the error.

**Comment by ibnesayeed** [comment]
It was interesting to find out that the record that was nothing but a stopword was paced at line 4499. We are using records 1-4000 for training and 4001-5000 for testing and it happened to be the second last record on the test set. This was coincidental, but a good one that made the test more reliable.

**Comment by Ch4s3** [comment]
yeah, that was a good catch
